pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,387 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	echo "github.com/labstack/echo/v4"
+	echomiddleware "github.com/labstack/echo/v4/middleware"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"go.thethings.network/lorawan-stack/pkg/web/middleware"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	errEmptyGatewayEUI = errors.Define("empty_gateway_eui", "empty gateway EUI")
+	errListener        = errors.DefineFailedPrecondition(
+		"listener",
+		"failed to serve Basic Station frontend listener",
+	)
+)
+
+type srv struct {
+	ctx       context.Context
+	server    io.Server
+	webServer *echo.Echo
+	upgrader  *websocket.Upgrader
+	tokens    io.DownlinkTokens
+}
+
+func (*srv) Protocol() string { return "basicstation" }
+
+// New creates the Basic Station front end.
+func New(ctx context.Context, server io.Server) *echo.Echo {
+	webServer := echo.New()
+	webServer.Logger = web.NewNoopLogger()
+	webServer.HTTPErrorHandler = errorHandler
+	webServer.Use(
+		middleware.ID(""),
+		echomiddleware.BodyLimit("16M"),
```

Comment:
This can definitely be smaller than 16M

Reply:
What do you recommend?

Reply:
I'm okay with leaving it the way it is for this PR, just saying that it could be smaller